### DISCUSSION TITLE
quay.io private repo with a proxy - auth failures

### DISCUSSION THREAD

**Discussion by lmilbaum** [renovatebot/renovate#21594] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitLab EE

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

Renovate to lookup the image in the self-hosted container registry

### What you tried so far.

The image names listed in the Dockerfile are without the registry name.

Added the following code to config.js:

hostRules: [
    {
      hostType: 'docker',
      matchHost: 'self-hosted.redhat.com',
    },
  ]

### Relevant debug logs

<details><summary>Logs</summary>

```
[log](https://pastebin.com/jmmttxNY)```

</details>

**Comment by rarkins** [comment]
The code mentioned above does nothing. What was the intention of it?

**Reply by rarkins** [reply]
And please try to elaborate, it doesn't help to say "my intention was for it to work" or "my intention was for it to look up dependencies" :)

**Reply by lmilbaum** [reply]
You are right. Sorry for that. Amending the description.

**Comment by rarkins** [comment]
When you try without any hostRules, what problem/failure occurs? E.g. the internal Docker registry requires authentication?

**Reply by lmilbaum** [reply]
Without hostRules, it tries to reach out the docker.io registry. The internal Docker registry doesn't require authentication because I am using internal gitlab runners.

**Reply by rarkins** [reply]
In that case I would have tried a package rule with matchDatasources=docker, registryUrls=your.internal.registry

**Reply by lmilbaum** [reply]
ok. Will try that.

**Reply by lmilbaum** [reply]
Same result. 
config.js
```module.exports = {
  platform: 'gitlab',
  onboarding: true,
  onboardingConfig: {
    extends: [
      "config:base",
      ":dependencyDashboard",
      ":enablePreCommit",
      "group:allNonMajor",
      "workarounds:supportRedHatImageVersion"
    ],
  },
  "automerge": false,
  repositories: [
    'renovate-poc/renovate',
    'renovate-poc/go-toolset-container'
  ],
  hostRules: [
    {
      hostType: 'docker',
      matchHost: 'https://registry-proxy.engineering.redhat.com',
    }
  ],
  packageRules: [
    {
      matchDatasources: 'docker',
      registryUrls: ['https://registry-proxy.engineering.redhat.com'],
    }
  ]
};

log file:
DEBUG: packageFiles with updates (repository=renovate-poc/go-toolset-container, baseBranch=rhel-9.2.0)
       "config": {
         "dockerfile": [
           {
             "deps": [
               {
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "currentValue": "1-344",
                 "datasource": "docker",
                 "depName": "ubi9/s2i-base",
                 "depType": "final",
                 "packageName": "ubi9/s2i-base",
                 "replaceString": "ubi9/s2i-base:1-344",
                 "versioning": "docker",
                 "warnings": [
                   {
                     "topic": "ubi9/s2i-base",
                     "message": "Failed to look up docker package ubi9/s2i-base"
                   }
                 ],
                 "updates": []
               }
             ],
             "packageFile": "Dockerfile"
           }
         ]
       }```

**Reply by lmilbaum** [reply]
Just a second. I was a bit hasty.

**Reply by viceice** [reply]
please format your code properly, it's not easy readable

**Reply by lmilbaum** [reply]
@viceice Thanks for you feedback. I was a bit hasty earlier today and that what came out of it. Will format my code from now on.

**Comment by lmilbaum** [comment]
After taking a deeper look, we use a proxy to quay.io.  Does it matter if we interface through a proxy instead of directly with quay.io?

**Reply by lmilbaum** [reply]
I have updated the config.js to use a Bearer token but according to the log, it is not picked up.
```DEBUG: GET https://<reducted>/v2/auth?service=&scope=repository:<reducted>:pull = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=67) (repository=renovate-poc/go-toolset-container, baseBranch=<reducted>)
DEBUG: Unauthorized docker lookup (repository=renovate-poc/go-toolset-container, baseBranch=<reducted>)
       "registryHost": "https://<reducted>.",
       "dockerRepository": "<reducted>"
DEBUG: Response code 401 (UNAUTHORIZED) (repository=renovate-poc/go-toolset-container, baseBranch=<reducted>)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1681928[225](https://<reducted>/renovate-poc/renovate/-/jobs/11763236#L225)076,
           "socket": 1681928225077,
           "lookup": 1681928225079,
           "connect": 1681928225090,
           "secureConnect": 1681928225109,
           "upload": 1681928225109,
           "response": 1681928225142,
           "end": 1681928225143,
           "phases": {
             "wait": 1,
             "dns": 2,
             "tcp": 11,
             "tls": 19,
             "request": 0,
             "firstByte": 33,
             "download": 1,
             "total": 67
           }
         },
         "message": "Response code 401 (UNAUTHORIZED)",
         "stack": "HTTPError: Response code 401 (UNAUTHORIZED)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/35.17.1/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/35.17.1 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "authorization": "***********",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://<reducted>//v2/auth?service=&scope=repository:<reducted>:pull",
           "hostType": "docker",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 401,
           "statusMessage": "UNAUTHORIZED",
           "body": {
             "errors": [
               {
                 "code": "UNAUTHORIZED",
                 "detail": {},
                 "message": "Specified credentials are invalid"
               }
             ]
           },
           "headers": {
             "server": "nginx/1.20.1",
             "date": "Wed, 19 Apr 2023 18:17:05 GMT",
             "content-type": "application/json",
             "content-length": "95",
             "www-authenticate": "Bearer realm=\"https://quay.io/v2/auth\",service=\"quay.io\"",
             "docker-distribution-api-version": "registry/2.0",
             "set-cookie": [
               "850e789cc33a49f3191618047b43fbf0=074d5b9e825ebd641028f96a6633e8b0; path=/; HttpOnly; Secure; SameSite=None"
             ],
             "connection": "close"
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
DEBUG: Failed to get authHeaders for getTags lookup (repository=renovate-poc/go-toolset-container, baseBranch=<reducted>)
DEBUG: Failed to look up docker package <reducted> (repository=renovate-poc/go-toolset-container, baseBranch=<reducted>, packageFile=Dockerfile, dependency=<reducted>)```

**Reply by lmilbaum** [reply]
The request headers might be off. I would expect to see the following header in the list:
`Authorization: Bearer $TOKEN`. It has `authorization`.

**Reply by lmilbaum** [reply]
The errors looks similar to the ones mentioned over https://github.com/renovatebot/renovate/discussions/10188

**Reply by rarkins** [reply]
Headers are case insensitive so `"authorization": "***********",` is not an indication of any problem. The error seems quite verbose:

```
           "statusCode": 401,
           "statusMessage": "UNAUTHORIZED",
           "body": {
             "errors": [
               {
                 "code": "UNAUTHORIZED",
                 "detail": {},
                 "message": "Specified credentials are invalid"
               }
             ]
           },
```

Have you configured a token using hostRules for quay.io?

**Reply by lmilbaum** [reply]
I can configure a Bearer token in a pre-renovate step with the proxy endpoint but using the proxy endpoint instead of quay.io changes the way renovate operates. So the token is not really working with the quay.io endpoint.
I am not aware of a way to generate a Bearer token with quay.io directly.

The other option is to generate [OAuth token of an organisation application](https://docs.quay.io/api/#generating-a-token-for-internal-application-use) which currently not an option. I am not the quay.io organization admin.

**Reply by viceice** [reply]
i think you need to configure the proxy via normal Linux proxy environment variables and then add a host rule for quay.io with the quay.io credentials. use username and password for docker like you would do with docker cli.

**Reply by lmilbaum** [reply]
Not sure this is an option. The proxy is configured by the quay admin folks in my organization. It provides an interface for bearer token generation.  
Re-the username/password option, already tried that with no avail. AFAIK, the quay.io api calls do not work if only username/password used.

**Reply by viceice** [reply]
how do you configure and use the proxy with the docker cli?

**Reply by lmilbaum** [reply]
I have a proxy endpoint which is used as the host url instead of quay.io.

**Reply by viceice** [reply]
this doesn't answer the question.

how do you authenticate with docker cli? 

how looks your image name when running via docker cli?

any other special docker daemon config?

**Reply by lmilbaum** [reply]
Ohh. Apologies. I didn't understand your initial question. The following script might shed some light on what is working.

```
#!/bin/bash -x
set -euo pipefail
 
 
REPO=<namespace/image_name>
TAG='1-344'
REGISTRY=<proxy registry endpoint to quay.io>
TYPE='manifests'
 
AUTH_API=$(
    curl -sIkL -H "Accept: application/vnd.docker.distribution.manifest.v1+json" \
        $REGISTRY/v2/$REPO/$TYPE/$TAG| \
        grep -i 'www-authenticate:' | grep -oP 'realm="\Khttps://\S+(?=")'
)
 
# Get bearer token
TOKEN=$(curl -sLk "${AUTH_API}?scope=repository:${REPO}:pull" | jq '.token' -r)

ALL_DIGEST=$(
    curl -skL \
        -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" \
        -H "Authorization: Bearer $TOKEN" \
        $REGISTRY/v2/$REPO/tags/list > ./all_dig) 
 
cat ./all_dig
```

**Reply by lmilbaum** [reply]
@viceice Our internal interface with the registry is quite complex.  There are all kind of interfaces and wrappers which works behind the hood which prevents me from figuring this out. I am aware that this is not a good enough answer but the script I provided in the previous comment is what I got.

**Reply by lmilbaum** [reply]
What I have noticed from the logs is that the first api call to get the Bearer token is probably not invoked and there for Bearer token is not used by the function `getTagsQuayRegistry`. As a result, it  throws `401`.  That is what happening when I use the `quay.io` endpoint. When I use the proxy endpoint, Renovate "thinks" that it is a docker registry. I've tried to use the alias with no avail.

**Reply by viceice** [reply]
renovate uses the quay.io API only for quay.io images, not for any other (proxy) docker registry. so renovate uses the default docker API.

your script doesn't send any token to the registry, so you should simply don't configure any hostrule for quay.io and your proxy.

**Reply by lmilbaum** [reply]
BTW,  I am able to login with my username/password credentials. 
```
[lmilbaum@lmilbaum renovate]$ podman login quay.io
Username: lmilbaum
Password: 
Login Succeeded!
```

**Reply by lmilbaum** [reply]
@viceice The proxy docker registry is not another one. It is just a front end to quay.io. 
The TOKEN is used in the last curl command.

**Reply by viceice** [reply]
yes, but you don't need any auth to get the token.

so it's the normal docker flow.
1. query tags list which returns 401 and the www-authenticate header
2. query the url from Header and get the token
3. query tags with bearer token again

did you tried renovate without a hostrule for the proxy?

**Reply by viceice** [reply]
> BTW,  I am able to login with my username/password credentials. 
> ```
> [lmilbaum@lmilbaum renovate]$ podman login quay.io
> Username: lmilbaum
> Password: 
> Login Succeeded!
> ```

this doesn't help. i asked you how do you login podman to your proxy.

**Reply by lmilbaum** [reply]
No need to handle the proxy. I can use quay.io directly. I have managed to make the script working with quay.io

```
#!/bin/bash -x
set -euo pipefail
 
 
REPO='<reducted>'
TAG='1-344'
REGISTRY='https://quay.io'
TYPE='manifests'
 

TOKEN=$(curl -sLk --user <quay user>:<quay encrypted cli password> "https://quay.io/v2/auth?scope=repository:${REPO}:pull&service=quay.io" | jq '.token' -r)



ALL_DIGEST=$(
    curl -skL \
        -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" \
        -H "Authorization: Bearer $TOKEN" \
        $REGISTRY/v2/$REPO/tags/list > ./all_dig)
 
 
cat ./all_dig
```

**Reply by lmilbaum** [reply]
BTW, the login to the proxy endpoint is the same as to quay.io directly. Same username/password.

**Reply by lmilbaum** [reply]
Here is how hostRules looks like:
```hostRules: [
    {
      "hostType": "docker",
      "matchHost": "https://quay.io/",
      "username": process.env.QUAY_USERNAME,
      "password": process.env.QAUY_PASSWORD
    }
  ],```

Here is a snippet from the log file (Running the renovate container on my laptop):

```
       "originalConfig": {
         "platform": "gitlab",
         "onboarding": true,
         "onboardingConfig": {
           "extends": [
             "config:base",
             ":dependencyDashboard",
             ":enablePreCommit",
             "group:allNonMajor",
             "workarounds:supportRedHatImageVersion"
           ]
         },
         "automerge": false,
         "repositories": ["<reducted>"],
         "hostRules": [
           {
             "hostType": "docker",
             "matchHost": "https://quay.io/",
             "username": "lmilbaum"
           }
         ],
         "packageRules": [{"matchDatasources": "docker", "registryUrls": ["quay.io"]}],
         "enabledManagers": ["dockerfile"]
       },
       "migratedConfig": {
         "platform": "gitlab",
         "onboarding": true,
         "onboardingConfig": {
           "extends": [
             "config:base",
             ":dependencyDashboard",
             ":enablePreCommit",
             "group:allNonMajor",
             "workarounds:supportRedHatImageVersion"
           ]
         },
         "automerge": false,
         "repositories": ["renovate-poc/go-toolset-container"],
         "hostRules": [
           {
             "hostType": "docker",
             "matchHost": "https://quay.io/",
             "username": "lmilbaum"
           }
         ],
         "packageRules": [{"matchDatasources": "docker", "registryUrls": ["quay.io"]}],
         "enabledManagers": ["dockerfile"]
       }```

The password is not there which is weired

**Reply by lmilbaum** [reply]
Ignore the password issue. I had a typo.

**Reply by lmilbaum** [reply]
Snippet from the log file:

```
TRACE: Applying Basic authentication (repository=<reducted>, baseBranch=<reducted>)
       "url": "https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true"
TRACE: No throttle (repository=<reducted>, baseBranch=<reducted>)
       "host": "quay.io"
TRACE: No concurrency limits (repository=renovate-poc/go-toolset-container, baseBranch=rhel-9.2.0)
       "host": "quay.io"
TRACE: got request (repository=<reducted>, baseBranch=<reducted>)
       "url": "https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "responseType": "json",
         "headers": {
           "accept": "application/json",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)",
           "authorization": "***********"
         },
         "hooks": {"beforeRedirect": ["[function]"]},
         "timeout": 60000
       }
DEBUG: GET https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=772) (repository=<reducted>, baseBranch=<reducted>)
DEBUG: Datasource unauthorized (repository=<reducted>, baseBranch=<reducted>)
       "datasource": "docker",
       "packageName": "<reducted>",
       "url": "https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true"
DEBUG: Failed to look up docker package <reducted> (repository=<reducted>, baseBranch=<reducted>, packageFile=Dockerfile, dependency=<reducted>)
TRACE: fetchManagerPackagerFileUpdates finished (repository=r<reducted>, baseBranch=<reducted>, packageFile=Dockerfile)
```

**Reply by lmilbaum** [reply]
Pay attention that renovate uses v1 api schema while the script I've shared is based on v2 api schema.

**Comment by lmilbaum** [comment]
A short recap of what I have tried so far:
1. hostRules - proxy and quay.io
2. packageRules - proxy and quay.io
3. registryAlias
4. adding username and password key/values

None of those options worked. Here is the latest log snippet:

```DEBUG: GET https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=79) (repository=<reducted>, baseBranch=<reducted>)
DEBUG: Datasource unauthorized (repository=<reducted>, baseBranch=<reducted>)
       "datasource": "docker",
       "packageName": "<reducted>",
       "url": "https://quay.io/api/v1/repository/<reducted>/tag/?limit=100&page=1&onlyActiveTags=true"
DEBUG: Failed to look up docker package <reducted> (repository=<reducted>, baseBranch=<reducted>, packageFile=Dockerfile, dependency=<reducted>)```

**Reply by lmilbaum** [reply]
The command I use for running renovate:
```[lmilbaum@lmilbaum renovate]$ docker run     --rm     -e LOG_LEVEL=trace  --privileged -e QUAY_USERNAME=lmilbaum -e QUAY_PASSWORD="<reducted>"   -e GIT_SSL_NO_VERIFY=1 -e NODE_TLS_REJECT_UNAUTHORIZED=0 -v ${PWD}/config.js:/usr/src/app/config.js     renovate/renovate:latest     --token "<reducted>"     --dry-run="true" > stam.log```

**Comment by viceice** [comment]
starting a new answer. so please try the following and paste http request from the logs for the proxy

1. create a repo
2. add a `Dockerfile` and a basic renovate.json
3. add a `FROM your.quay.proxy/some/image` to the dockerfile (replace with real existing)
4. run renovate
5. provide full debug logs


don't provide any hostrule for the proxy

**Reply by lmilbaum** [reply]
Dockerfile
```
FROM fake_group/fake_repo:1-1
```

log file
[stam.log](https://github.com/renovatebot/renovate/files/11293501/stam.log)

**Reply by lmilbaum** [reply]
Added a packageRule for the proxy

[stam.log](https://github.com/renovatebot/renovate/files/11293637/stam.log)

**Reply by viceice** [reply]
```
"https://<proxy quay.io server>/v2/auth (GET,200)": 1,
         "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list (GET,401)": 2
```

ok. renovate is first calling the tag list. then it gets the 401 and is calling the auth endpoint from the www-authenticate header.
it also seems to get a token, but that doesn't seem to be accepted by the tags list API. 🤔

**Reply by viceice** [reply]
can you run a trace run and snip the log parts where the docker API calls are made?

**Reply by lmilbaum** [reply]
```
TRACE: Cache miss (repository=lmilbaum/renovate_quay)
       "namespace": "datasource-docker-tags",
       "key": "cache-decorator:<proxy quay.io server>:fake_group/fake_repo"
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
TRACE: No throttle (repository=lmilbaum/renovate_quay)
       "host": "<proxy quay.io server>"
TRACE: No concurrency limits (repository=lmilbaum/renovate_quay)
       "host": "<proxy quay.io server>"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "throwHttpErrors": false,
         "noAuth": true,
         "responseType": "json",
         "headers": {
           "accept": "application/json",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "hooks": {"beforeRedirect": ["[function]"]},
         "timeout": 60000
       }
TRACE: Obtaining docker registry token (repository=lmilbaum/renovate_quay)
       "registryHost": "<proxy quay.io server>",
       "dockerRepository": "fake_group/fake_repo",
       "authUrl": "<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull"
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "timeout": 60000,
         "noAuth": true,
         "responseType": "json",
         "headers": {
           "accept": "application/json",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "hooks": {"beforeRedirect": ["[function]"]}
       }
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "headers": {
           "accept": "application/json",
           "authorization": "***********",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "noAuth": true,
         "responseType": "json",
         "hooks": {"beforeRedirect": ["[function]"]},
         "timeout": 60000
       }
DEBUG: GET <proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=508) (repository=lmilbaum/renovate_quay)
DEBUG: Datasource unauthorized (repository=lmilbaum/renovate_quay)
       "datasource": "docker",
       "packageName": "fake_group/fake_repo",
       "url": "<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
DEBUG: Failed to look up docker package fake_group/fake_repo (repository=lmilbaum/renovate_quay, packageFile=Dockerfile, dependency=fake_group/fake_repo)
```

**Reply by lmilbaum** [reply]
`https://<proxy quay.io server>/v2/auth` should be provided with username and password to return a token

**Reply by lmilbaum** [reply]
And the service should be `quay.io`.  Similar to what this command is doing:
```
TOKEN=$(curl -sLk --user <username>:<password> "https://<proxy registry url/v2/auth?scope=repository:${REPO}:pull&service=quay.io" | jq '.token' -r)

```

I've checked if proxy can be used instead and it is not. Only quay.io.

**Reply by viceice** [reply]
> `https://<proxy quay.io server>/v2/auth` should be provided with username and password to return a token

your sample shell scripts don't use credentials

**Reply by viceice** [reply]
> And the service should be `quay.io`.  Similar to what this command is doing:
> ```
> TOKEN=$(curl -sLk --user <username>:<password> "https://<proxy registry url/v2/auth?scope=repository:${REPO}:pull&service=quay.io" | jq '.token' -r)
> 
> ```

ok, for this you should now configure a hostrule for the auth url with username and password (even a token needs to be passed as password like you would do for docker or podman cli)


> 
> I've checked if proxy can be used instead and it is not. Only quay.io.

service should come from the www-authenticate header 🤔

can you login with podman to the proxy and successfully pull an image from it?

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/bf0e39da5fc7f4e7aec16ef95b15e421426537aa/lib/modules/datasource/docker/index.ts#L163-L166

here we set an empty service if not found on www-authenticate header

**Reply by lmilbaum** [reply]
> > And the service should be `quay.io`.  Similar to what this command is doing:
> > ```
> > TOKEN=$(curl -sLk --user <username>:<password> "https://<proxy registry url/v2/auth?scope=repository:${REPO}:pull&service=quay.io" | jq '.token' -r)
> > ```
> 
> ok, for this you should now configure a hostrule for the auth url with username and password (even a token needs to be passed as password like you would do for docker or podman cli)
> 
> > I've checked if proxy can be used instead and it is not. Only quay.io.
> 
> service should come from the www-authenticate header thinking
> 
> can you login with podman to the proxy and successfully pull an image from it?

Yes

**Reply by lmilbaum** [reply]
With hostRules:
```
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in /usr/src/app/config.js
 WARN: cli config dryRun property has been changed to full
TRACE: mergeChildConfig
       "parent": {
         "platform": "gitlab",
         "endpoint": "<self hosted gitlab>",
         "repositories": ["lmilbaum/renovate_quay"],
         "hostRules": [
           {
             "hostType": "docker",
             "matchHost": "<proxy>",
             "username": "lmilbaum",
             "password": "***********"
           }
         ],
         "packageRules": [
           {
             "matchDatasources": "docker",
             "registryUrls": ["<proxy>"]
           }
         ],
         "enabledManagers": ["dockerfile"]
       },
       "child": {"hostRules": []}
```

**Reply by lmilbaum** [reply]
```TRACE: Cache miss (repository=lmilbaum/renovate_quay)
       "namespace": "datasource-docker-tags",
       "key": "cache-decorator:https://<proxy quay.io server>:fake_group/fake_repo"
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
TRACE: No throttle (repository=lmilbaum/renovate_quay)
       "host": "<proxy quay.io server>"
TRACE: No concurrency limits (repository=lmilbaum/renovate_quay)
       "host": "<proxy quay.io server>"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "throwHttpErrors": false,
         "noAuth": true,
         "responseType": "json",
         "headers": {
           "accept": "application/json",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "hooks": {"beforeRedirect": ["[function]"]},
         "timeout": 60000
       }
TRACE: Using basic auth for Docker registry (repository=lmilbaum/renovate_quay)
       "registryHost": "https://<proxy quay.io server>",
       "dockerRepository": "fake_group/fake_repo"
TRACE: Obtaining docker registry token (repository=lmilbaum/renovate_quay)
       "registryHost": "https://<proxy quay.io server>",
       "dockerRepository": "fake_group/fake_repo",
       "authUrl": "https://<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull"
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/auth?service=&scope=repository:fake_group/fake_repo:pull",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "timeout": 60000,
         "headers": {
           "accept": "application/json",
           "authorization": "***********",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "noAuth": true,
         "responseType": "json",
         "hooks": {"beforeRedirect": ["[function]"]}
       }
TRACE: Authorization disabled (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
TRACE: got request (repository=lmilbaum/renovate_quay)
       "url": "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "hostType": "docker",
         "headers": {
           "accept": "application/json",
           "authorization": "***********",
           "user-agent": "RenovateBot/35.54.0 (https://github.com/renovatebot/renovate)"
         },
         "noAuth": true,
         "responseType": "json",
         "hooks": {"beforeRedirect": ["[function]"]},
         "timeout": 60000
       }
DEBUG: GET https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=575) (repository=lmilbaum/renovate_quay)
DEBUG: Datasource unauthorized (repository=lmilbaum/renovate_quay)
       "datasource": "docker",
       "packageName": "fake_group/fake_repo",
       "url": "https://<proxy quay.io server>/v2/fake_group/fake_repo/tags/list?n=10000"
DEBUG: Failed to look up docker package fake_group/fake_repo (repository=lmilbaum/renovate_quay, packageFile=Dockerfile, dependency=fake_group/fake_repo)
```

**Reply by lmilbaum** [reply]
Using proxy
```
[lmilbaum@lmilbaum renovate]$ curl -skLi -H 'Accept: application/vnd.docker.distribution.manifest.list.v2+json' https://<proxy>/v2/<fake_group>/<fake_repo>/tags/list
HTTP/2 401 
server: nginx/1.20.1
date: Fri, 21 Apr 2023 12:14:41 GMT
content-type: application/json
content-length: 112
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://<proxy>/v2/auth"
set-cookie: 850e789cc33a49f3191618047b43fbf0=4267313853ad6a1c1a19f7f848ac3c64; path=/; HttpOnly; Secure; SameSite=None
```

**Reply by lmilbaum** [reply]
For comparison using quay.io
```
[lmilbaum@lmilbaum renovate]$ curl -skLi -H 'Accept: application/vnd.docker.distribution.manifest.list.v2+json' https://quay.io/v2/<fake_group>/<fake_repo>/tags/list
HTTP/2 401 
date: Fri, 21 Apr 2023 12:17:48 GMT
content-type: application/json
content-length: 112
server: nginx/1.20.1
www-authenticate: Bearer realm="https://quay.io/v2/auth",service="quay.io",scope="repository:<fake_group>/<fake_repo>:pull"
docker-distribution-api-version: registry/2.0

{"errors":[{"code":"UNAUTHORIZED","detail":{},"message":"access to the requested resource is not authorized"}]}
```

**Reply by lmilbaum** [reply]
@viceice Is #11449 related in anyway?

**Reply by lmilbaum** [reply]
Something is off. I thought that the empty `service` is the reason the auth command is not returning a token but when running the following command, it does return a token.
```
curl https://<proxy>/v2/auth?service=&scope=repository:<namespace:image>:pull
```

**Reply by viceice** [reply]
> @viceice Is #11449 related in anyway?

no

**Reply by lmilbaum** [reply]
I've added some printing information and nailed the exact error returned from the second try to get the tags

```
         "response": {
           "statusCode": 401,
           "statusMessage": "UNAUTHORIZED",
           "body": {"error": "Token is missing the \"aud\" claim"},
```

**Reply by lmilbaum** [reply, marked as answer]
I have the root cause of this problem. It related to #17294. The proxy does not return a `service` with the `www-authenticate` header. 
Returning an empty service query param back generates a token with 'aud'=''  while removing the service from the query param generates a token with 'aud'='quay.io'.  The one which is populated with quay.io is working while the empty one generates the error mentioned in the previous comment.

**Reply by viceice** [reply]
so we should completely remove service if it's empty instead of passing it with empty value?
